onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+  onnxruntime::Model model(graph.Name(), true, ModelMetaData(), IOnnxRuntimeOpSchemaRegistryList(), graph.DomainToVersionMap());
+  onnxruntime::Graph& graph_build = model.MainGraph();
+  const std::vector<NodeIndex>& node_index = graph.GetNodesInTopologicalOrder();
+  for (const auto& index : node_index) {
+    const Node* node = graph.GetNode(index);
+    graph_build.AddNode(*node);
+  }
+  ORT_ENFORCE(graph_build.Resolve().IsOK());
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  // Serialize modelproto to string
+  string string_buf;
+  model_proto.SerializeToString(&string_buf);
+
+  // Get supported node list
+  SubGraphCollection_t supported_nodes_vector;
+  TRTLogger trt_logger(static_cast<nvinfer1::ILogger::Severity>(nvinfer1::ILogger::Severity::kWARNING));
+  auto trt_builder = unique_pointer<nvinfer1::IBuilder>(nvinfer1::createInferBuilder(trt_logger));
+  auto trt_network = unique_pointer<nvinfer1::INetworkDefinition>(trt_builder->createNetwork());
+  auto trt_parser  = unique_pointer<nvonnxparser::IParser>(nvonnxparser::createParser(trt_network.get(), trt_logger));
+  trt_parser->supportsModel(string_buf.data(), string_buf.size(), supported_nodes_vector);
+  model_proto.release_graph();
+
+  // Find inputs, initializers and outputs for each supported subgraph
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  for (const auto& group : supported_nodes_vector) {
+    std::set<size_t> node_set(group.begin(), group.end());
+    std::unique_ptr<IndexedSubGraph> sub_graph = std::make_unique<IndexedSubGraph>();
+
+    if (!group.empty()) {
+      // Find inputs and outputs of the subgraph
+      std::map<const NodeArg*, int> fused_inputs, fused_outputs, fused_outputs_to_add;
+      std::set<const NodeArg*> erased;
+      int input_order = 0;
+      int output_order = 0;
+
+      for (const auto& index : group) {
+        sub_graph->nodes.push_back(index);
+        const auto& node = graph.GetNode(index);
+
+        for (const auto& input : node->InputDefs()) {
+          const auto& it = fused_outputs.find(input);
+
+          if (it != fused_outputs.end()) {
+            fused_outputs.erase(it);
+            erased.insert(input);
+          }
+          //only when input is neither in output list nor erased list, add the input to input list
+          else if (erased.find(input) == erased.end()) {
+            fused_inputs[input] = input_order++;
+          }
+        }
+
+        // For output searching, there is a special case:
+        // If node's OutputEdges are more than its outputs, meaning certain output is used more than once,
+        // if the output is connected to nodes that don't belong to the subgraph, the output need to be added
+        // to the output list
+        if (node->GetOutputEdgesCount() > node->OutputDefs().size()) {
+          for (auto it = node->OutputEdgesBegin(); it != node->OutputEdgesEnd(); ++it) {
```

Comment:
>node->OutputEdgesEnd() [](start = 57, length = 22)

Call OutputEdgesEnd() once outside the loop instead of on every iteration. 

You only want to do the call in the condition statement if the value may change during iteration.

Reply:
It's corrected. Thanks.